pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -429,11 +440,12 @@ public synchronized void readEntriesComplete(List<Entry> entries, Object ctx) {
                 // remove positions first from replay list first : sendMessages recycles entries
                 if (readType == ReadType.Replay) {
                     entries.subList(start, start + messagesForC).forEach(entry -> {
-                        messagesToReplay.remove(entry.getLedgerId(), entry.getEntryId());
+                        messagesToRedeliver.remove(entry.getLedgerId(), entry.getEntryId());
                     });
                 }
 
-                SendMessageInfo sentMsgInfo = c.sendMessages(entries.subList(start, start + messagesForC));
+                SendMessageInfo sentMsgInfo = c
+                        .sendMessages(new ArrayList<>(entries.subList(start, start + messagesForC)));
```

Comment:
can you explain why you are constructing a new array list?

Reply:
The mutation was also possible to happen before (eg: messages with wrong checksum were removed from the list)

Reply:
So why the new array list?